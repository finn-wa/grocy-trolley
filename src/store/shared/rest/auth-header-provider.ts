import { AppTokens } from "@gt/app/di";
import { LoginDetails } from "@gt/store/shared/rest/login-details.model";
import {
  CacheService,
  CacheServiceFactory,
  getCacheDir,
  sanitiseEmailForCache,
} from "@gt/utils/cache";
import { HeadersBuilder, headersFromRaw, headersToRaw } from "@gt/utils/headers";
import { Logger } from "@gt/utils/logger";
import { existsSync } from "fs";
import path from "path";
import { Browser, BrowserContext, Page } from "playwright";
import { inject } from "tsyringe";

/**
 * Provides authenticated headers for API requests to an online store. Logs in
 * to the store using Playwright and caches the headers.
 */
export abstract class AuthHeaderProvider {
  /**
   * Defines the headers to use for future requests and the headers to discard.
   */
  protected abstract readonly headersFilter: { allowed: string[]; disallowed: string[] };
  protected abstract readonly logger: Logger;

  private readonly headersCache: CacheService<{ headers: Record<string, string[]> }>;
  private browserState: BrowserState | null = null;
  /** Authenticated headers to use for API requests */
  private _authHeaders: Headers | null = null;

  /**
   * Creates a new AuthHeaderService.
   * @param name Name of store/website. Used for cache path.
   * @param browserLoader Cold promise that returns the Playwright Browser
   *    instance to use to perform requests.
   * @param loginDetails Optional login details. If provided, Playwright logs
   *    into this account and sends requests with credentials.
   */
  constructor(
    @inject(AppTokens.cacheServiceFactory)
    getCacheService: CacheServiceFactory<{ headers: Record<string, string[]> }>,
    public readonly name: string,
    protected readonly browserLoader: () => Promise<Browser>,
    protected readonly loginDetails?: LoginDetails | null
  ) {
    // use provider
    this.headersCache = getCacheService(
      path.join(name, sanitiseEmailForCache(loginDetails?.email))
    );
  }

  /**
   * Initialises the page to a logged-in state and return headers for future
   * authenticated requests. getHeadersFromRequest (in playwright utils) may be
   * helpful. Do not worry about filtering the headers, that will be done
   * automatically (remember to define {@link headersFilter}).
   * @param page fresh browser page to use to log in to website. Will be kept
   *              open for future requests if needed, so do not clean up.
   * @return captured authenticated headers
   */
  protected abstract login(page: Page): Promise<Headers>;

  /**
   * Used to test whether cached headers are valid or not.
   * @param headers Headers to test
   * @returns True if the headers are valid
   */
  protected abstract isValid(headers: Headers): Promise<boolean>;

  /**
   *
   * @returns Headers with auth information
   */
  async authHeaders(): Promise<HeadersBuilder> {
    if (!this._authHeaders) {
      this._authHeaders = await this.retrieveAuthHeaders();
    }
    return new HeadersBuilder(this._authHeaders);
  }

  private async retrieveAuthHeaders(): Promise<Headers> {
    this.logger.debug("Retrieving auth headers...");
    console.log(this.headersCache);

    const rawCachedHeaders = await this.headersCache.get("headers");
    if (rawCachedHeaders) {
      const cachedHeaders = headersFromRaw(rawCachedHeaders);
      if (await this.isValid(cachedHeaders)) {
        this.logger.debug("Using cached auth headers");
        this._authHeaders = cachedHeaders;
        return cachedHeaders;
      }
    }
    this.logger.info("No valid cached headers found, logging in...");
    this.browserState = await this.newBrowserState();
    const capturedHeaders = await this.login(this.browserState.page);
    const authHeaders = await this.filterCapturedHeaders(capturedHeaders);
    // Cache browser state post-login
    await this.browserState.context.storageState({ path: this.getPlaywrightStoragePath() });
    // Cache filtered headers
    await this.headersCache.set("headers", headersToRaw(authHeaders));
    this._authHeaders = authHeaders;
    return authHeaders;
  }

  /**
   * Returns a new browser context with cached storage state (if found).
   * This is a pure function that does not set fields.
   * @returns Browser context
   */
  private async newBrowserState(): Promise<BrowserState> {
    const browser = await this.browserLoader();
    const storageState = this.getPlaywrightStoragePath();
    const context = existsSync(storageState)
      ? await browser.newContext({ storageState })
      : await browser.newContext();
    return { context, page: await context.newPage() };
  }

  private getPlaywrightStoragePath(): string {
    return path.join(
      getCacheDir(),
      this.name,
      sanitiseEmailForCache(this.loginDetails?.email),
      "playwright.json"
    );
  }

  /**
   * Filters out disallowed headers from a Headers object.
   * @param capturedHeaders Headers captured from authenticated request in {@link login}
   * @returns New Headers object containing only permitted headers
   */
  private async filterCapturedHeaders(capturedHeaders: Headers): Promise<Headers> {
    const allowed = this.headersFilter.allowed.map((name) => name.toLowerCase());
    const disallowed = this.headersFilter.disallowed.map((name) => name.toLowerCase());
    const authHeaders = new Headers();
    for (const name of capturedHeaders.keys()) {
      const lowercaseName = name.toLowerCase();
      if (disallowed.includes(lowercaseName)) {
        continue;
      }
      if (!allowed.includes(lowercaseName)) {
        this.logger.warn(`Unrecognised header:\n${name}: ${capturedHeaders.get(name)}`);
      }
      authHeaders.set(name, capturedHeaders.get(name)!);
    }
    return authHeaders;
  }
}

interface BrowserState {
  context: BrowserContext;
  page: Page;
}
