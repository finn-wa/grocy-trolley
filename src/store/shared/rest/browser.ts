import { Browser, firefox } from "playwright-firefox";
import { LogLevel, playwrightLogger } from "@gt/utils/logger";

export let browser: Browser | null = null;

/**
 * Returns a lazy-loaded shared browser instance.
 * @returns The browser instance
 */
export async function getBrowser(): Promise<Browser> {
  if (browser && browser.isConnected()) {
    return browser;
  }
  browser = await firefox.launch({
    headless: false,
    logger: playwrightLogger(LogLevel.WARN),
  });
  return browser;
}
