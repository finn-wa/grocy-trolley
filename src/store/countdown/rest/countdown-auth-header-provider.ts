import { AppTokens } from "@gt/app/di";
import { AuthHeaderProvider } from "@gt/store/shared/rest/auth-header-provider";
import { LoginDetails } from "@gt/store/shared/rest/login-details.model";
import { CacheServiceFactory } from "@gt/utils/cache";
import { Logger } from "@gt/utils/logger";
import { getHeadersFromRequest } from "@gt/utils/playwright";
import { Browser, Page, Request as PlaywrightRequest } from "playwright";
import { inject, Lifecycle, scoped } from "tsyringe";
import { CountdownTokens } from "../countdown-di";
import { COUNTDOWN_URL, Shell } from "../models";

@scoped(Lifecycle.ContainerScoped)
export class CountdownAuthHeaderProvider extends AuthHeaderProvider {
  protected readonly logger = new Logger(this.constructor.name);
  protected readonly headersFilter = {
    allowed: [
      "host",
      "user-agent",
      "accept",
      "accept-language",
      "accept-encoding",
      "referer",
      "dnt",
      "cookie",
      "sec-fetch-dest",
      "sec-fetch-mode",
      "sec-fetch-site",
    ],
    disallowed: [
      "accept",
      "content-type",
      "x-newrelic-id",
      "newrelic",
      "traceparent",
      "tracestate",
      "__requestverificationtoken",
      "connection",
      "pragma",
      "cache-control",
      "te",
    ],
  };

  constructor(
    @inject(AppTokens.cacheServiceFactory)
    cacheServiceFactory: CacheServiceFactory<{ headers: Record<string, string[]> }>,
    @inject(AppTokens.browserLoader) browserLoader: () => Promise<Browser>,
    @inject(CountdownTokens.loginDetails) loginDetails?: LoginDetails
  ) {
    super(cacheServiceFactory, "countdown", browserLoader, loginDetails);
  }

  async login(page: Page): Promise<Headers> {
    const [_, shellRequest] = await Promise.all([
      page.goto(COUNTDOWN_URL + "/shop/securelogin"),
      page.waitForRequest(`${COUNTDOWN_URL}/api/v1/shell`),
    ]);
    if (!this.loginDetails || (await this.isLoggedIn(shellRequest))) {
      return getHeadersFromRequest(shellRequest);
    }
    await page.hover("global-nav-header");
    await page.waitForLoadState("networkidle");
    await page.type("#loginContainer input#loginID", this.loginDetails.email);
    await page.type("#loginContainer input#password", this.loginDetails.password);
    await page.click("#loginContainer input#remember");
    await page.click('#loginContainer button[value="Submit"]');
    // Arbitrary authenticated request that happens after login
    const request = await page.waitForRequest(`${COUNTDOWN_URL}/api/v1/shoppers/my/recipes`);
    return getHeadersFromRequest(request);
  }

  private async isLoggedIn(shellRequest: PlaywrightRequest): Promise<boolean> {
    const body = (await shellRequest.response().then((res) => res?.json())) as Shell;
    return body?.context?.shopper?.isLoggedIn;
  }

  protected async isValid(headers: Headers): Promise<boolean> {
    this.logger.warn("CountdownAuthHeaderProvider.isValid() is hardcoded to true");
    return true;
    // TODO this is returning false negatives
    /*
    const builder = new HeadersBuilder(headers).acceptJson();
    const response = await this.get(this.buildUrl("/v1/trolleys/my"), builder.build());
    if (!response.ok) return false;
    const body = await response.json();
    return body.isSuccessful;
    */
  }
}
