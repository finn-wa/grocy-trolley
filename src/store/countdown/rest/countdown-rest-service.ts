import { RestService } from "@gt/utils/rest";
import { COUNTDOWN_URL } from "../models";
import { CountdownAuthHeaderProvider } from "./countdown-auth-header-provider";

export abstract class CountdownRestService extends RestService {
  protected readonly baseUrl = this.validateBaseUrl(`${COUNTDOWN_URL}/api`);

  constructor(protected readonly authHeaderProvider: CountdownAuthHeaderProvider) {
    super();
  }

  protected authHeaders() {
    return this.authHeaderProvider.authHeaders();
  }
}
