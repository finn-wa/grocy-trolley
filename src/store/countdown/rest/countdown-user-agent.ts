import { StoreUserAgent } from "@gt/store/shared/rest/store-user-agent";
import { Logger } from "@gt/utils/logger";
import { BrowserContext, Page, Request as PlaywrightRequest } from "playwright-firefox";
import { COUNTDOWN_URL, Shell } from "../models";

export class CountdownUserAgent extends StoreUserAgent {
  public readonly storeName = "countdown";
  protected readonly logger = new Logger(this.constructor.name);

  async init(context: BrowserContext): Promise<{ page: Page; headers: Headers }> {
    const page = await context.newPage();
    const [_goto, shellRequest] = await Promise.all([
      page.goto(COUNTDOWN_URL + "/shop/securelogin"),
      page.waitForRequest(`${COUNTDOWN_URL}/api/v1/shell`),
    ]);
    if (!this.loginDetails || (await this.isLoggedIn(shellRequest))) {
      return { page, headers: await this.getHeadersFromRequest(shellRequest) };
    }
    await page.waitForLoadState("networkidle");
    await page.type("#loginContainer input#loginID", this.loginDetails.email);
    await page.type("#loginContainer input#password", this.loginDetails.password);
    await page.click("#loginContainer input#remember");
    await page.click('#loginContainer button[value="Submit"]');
    // Arbitrary authenticated request that happens after login
    const request = await page.waitForRequest(`${COUNTDOWN_URL}/api/v1/shoppers/my/recipes`);
    return { page, headers: await this.getHeadersFromRequest(request) };
  }

  private async isLoggedIn(shellRequest: PlaywrightRequest): Promise<boolean> {
    const body = (await shellRequest.response().then((res) => res?.json())) as Shell;
    return body?.context?.shopper?.isLoggedIn;
  }
}
