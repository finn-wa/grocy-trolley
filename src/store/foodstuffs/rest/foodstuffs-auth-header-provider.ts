import { AppTokens } from "@gt/app/di";
import { AuthHeaderProvider } from "@gt/store/shared/rest/auth-header-provider";
import { LoginDetails } from "@gt/store/shared/rest/login-details.model";
import { CacheServiceFactory } from "@gt/utils/cache";
import { HeadersBuilder } from "@gt/utils/headers";
import { Logger, prettyPrint } from "@gt/utils/logger";
import { getHeadersFromRequest } from "@gt/utils/playwright";
import { RestService } from "@gt/utils/rest";
import { Browser, Page } from "playwright";
import { inject, Lifecycle, scoped } from "tsyringe";
import { FoodstuffsTokens } from "../foodstuffs-di";
import { List } from "../lists/foodstuffs-list.model";
import { PAKNSAVE_URL } from "../models";

class FoodstuffsAuthHeaderTestService extends RestService {
  protected readonly baseUrl = `${PAKNSAVE_URL}/CommonApi`;
  protected readonly logger = new Logger(this.constructor.name);

  constructor(private readonly headers: Headers) {
    super();
  }

  getCart() {
    return this.getAndParse(this.buildUrl("/Cart/Index"), {
      headers: new HeadersBuilder(this.headers).acceptJson().build(),
    });
  }

  async createList(name: string): Promise<List> {
    return this.putAndParse(this.buildUrl("ShoppingLists/CreateList", { name }), {
      headers: new HeadersBuilder(this.headers).acceptJson().build(),
    });
  }

  async deleteList(id: string | number): Promise<Response> {
    return this.delete(this.buildUrl(`ShoppingLists/DeleteList/${id}`), {
      headers: new HeadersBuilder(this.headers).acceptJson().build(),
    });
  }
}

/**
 * Uses Playwright to perform Foodstuffs requests from a browser. Necessary
 * because Cloudflare now blocks requests that are not sent from a browser.
 */
@scoped(Lifecycle.ContainerScoped)
export class FoodstuffsAuthHeaderProvider extends AuthHeaderProvider {
  protected readonly logger = new Logger(this.constructor.name);
  protected readonly headersFilter = {
    allowed: [
      "host",
      "user-agent",
      "accept-language",
      "accept-encoding",
      "referer",
      "dnt",
      "cookie",
      "sec-fetch-dest",
      "sec-fetch-mode",
      "sec-fetch-site",
    ],
    disallowed: [
      "accept",
      "content-type",
      "x-newrelic-id",
      "newrelic",
      "traceparent",
      "tracestate",
      "__requestverificationtoken",
      "connection",
      "pragma",
      "cache-control",
      "te",
    ],
  };

  constructor(
    @inject(AppTokens.cacheServiceFactory)
    cacheServiceFactory: CacheServiceFactory<{ headers: Record<string, string[]> }>,
    @inject("BrowserLoader") browserLoader: () => Promise<Browser>,
    @inject(FoodstuffsTokens.loginDetails) loginDetails?: LoginDetails
  ) {
    super(cacheServiceFactory, "foodstuffs", browserLoader, loginDetails);
  }

  async login(page: Page): Promise<Headers> {
    await page.goto(`${PAKNSAVE_URL}/shop`);
    let cartRequest = await page.waitForRequest(`${PAKNSAVE_URL}/CommonApi/Cart/Index`);

    if (this.loginDetails && !(await this.isLoggedIn(page))) {
      this.logger.info("Logging in to foodstuffs...");
      await page.click('button[id="login-form"]');
      await page.fill('input[id="login-email"]', this.loginDetails.email);
      await page.fill('input[id="login-password"]', this.loginDetails.password);
      await page.click("button.login-form-submit");
      cartRequest = await page.waitForRequest(`${PAKNSAVE_URL}/CommonApi/Cart/Index`);
      // page refreshes after submission
      await page.waitForLoadState("networkidle");
      // search box seems to pop in last
      await page.locator('input[aria-label="Search products"]').waitFor();
    }
    return getHeadersFromRequest(cartRequest);
  }

  private async isLoggedIn(page: Page) {
    type DataLayer = { loginState?: "loggedIn" | "guest" }[];
    const loginState = await page.evaluate(() => {
      const dataLayer = (globalThis as unknown as { dataLayer: DataLayer }).dataLayer;
      return dataLayer.find((entry) => "loginState" in entry)?.loginState;
    });
    return loginState === "loggedIn";
  }

  protected async isValid(headers: Headers): Promise<boolean> {
    const testService = new FoodstuffsAuthHeaderTestService(headers);
    try {
      const cart = await testService.getCart();
      if (!cart || typeof cart !== "object" || Object.keys(cart).length === 0) {
        // We expect getCart to throw an error if it failed, but if it returns ok check the body
        throw new Error("Test getCart request failed: " + prettyPrint(cart));
      }
      const list = await testService.createList("[temp]");
      if (
        !list ||
        typeof list !== "object" ||
        Object.keys(list).length === 0 ||
        !("listId" in list)
      ) {
        throw new Error("Test createList request failed: " + prettyPrint(cart));
      }
      await testService.deleteList(list.listId);
    } catch (error) {
      this.logger.debug("Headers failed validation", headers, error);
      return false;
    }
    return true;
  }
}

export interface FoodstuffsUserProfile {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  agreedToMarketing: boolean;
  agreedToTermsAndConditions: boolean;
  clubCardUser: boolean;
}

export interface LoginResponse {
  success: boolean;
  userProfile: FoodstuffsUserProfile;
}
