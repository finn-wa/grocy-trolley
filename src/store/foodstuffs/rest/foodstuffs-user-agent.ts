import { LoginDetails } from "@gt/store/shared/rest/login-details.model";
import { StoreUserAgent } from "@gt/store/shared/rest/store-user-agent";
import { APPLICATION_JSON, headersToRaw } from "@gt/utils/headers";
import { Logger, prettyPrint } from "@gt/utils/logger";
import { BrowserContext, JSHandle, Page } from "playwright-firefox";
import { PAKNSAVE_URL } from "../models";

/**
 * Uses Playwright to perform Foodstuffs requests from a browser. Necessary
 * because Cloudflare now blocks requests that are not sent from a browser.
 */
export class FoodstuffsUserAgent extends StoreUserAgent {
  public readonly storeName = "foodstuffs";
  protected readonly logger = new Logger(this.constructor.name);

  async init(context: BrowserContext): Promise<{ page: Page; headers: Headers }> {
    const page = await context.newPage();
    await page.goto(`${PAKNSAVE_URL}/shop`);
    let cartRequest = await page.waitForRequest(`${PAKNSAVE_URL}/CommonApi/Cart/Index`);

    if (this.loginDetails && !(await this.isLoggedIn(page))) {
      this.logger.info("Logging in to foodstuffs...");
      await page.click('button[id="login-form"]');
      await page.fill('input[id="login-email"]', this.loginDetails.email);
      await page.fill('input[id="login-password"]', this.loginDetails.password);
      await page.click("button.login-form-submit");
      cartRequest = await page.waitForRequest(`${PAKNSAVE_URL}/CommonApi/Cart/Index`);
      // page refreshes after submission
      await page.waitForLoadState("networkidle");
      // search box seems to pop in last
      await page.locator('input[aria-label="Search products"]').waitFor();
    }
    const headers = await this.getHeadersFromRequest(cartRequest);
    return { page, headers };
  }

  /**
   * Creates a new agent with the same browser but new login details
   * @param loginDetails Login details for the new agent, or null to omit them
   * @returns the new agent
   */
  clone(loginDetails: LoginDetails | null): FoodstuffsUserAgent {
    return new FoodstuffsUserAgent(this.browserLoader, loginDetails ?? undefined);
  }

  /**
   * Performs a fetch request as a logged in user on the PAK'n'SAVE website.
   * @param method Request method
   * @param url Request URL
   * @param headers Request headers
   * @param body Request body
   * @returns Request response
   */
  async fetchWithBrowser(
    method: string,
    url: string,
    headers?: Headers,
    body?: BodyInit
  ): Promise<Response> {
    const page = await this.getLoginPage();
    const contentType = headers?.get("content-type");
    if (contentType === APPLICATION_JSON && body) {
      body = JSON.stringify(body);
    }
    const rawHeaders = headers ? headersToRaw(headers) : undefined;
    const responseHandle = await page.evaluateHandle(
      async ({ url, method, rawHeaders, body }) =>
        fetch(url, {
          credentials: this.loginDetails ? "include" : undefined,
          referrer: "https://www.paknsave.co.nz/shop",
          mode: "cors",
          method,
          // Record<string, string[]> is not in HeadersInit but it works in firefox at least
          headers: rawHeaders as unknown as HeadersInit | undefined,
          body,
        }),
      { url, method, rawHeaders, body }
    );
    // Response is not serialisable, so we must serialise it
    /* eslint-disable */
    const response = await responseHandle.evaluate((response, props) => {
      const responseJson: Record<string, any> = Object.fromEntries(
        props.map((key) => [key, response[key]])
      );
      responseJson.headers = Object.fromEntries(responseJson.headers.entries());
      return responseJson as SerialisedResponse;
    }, SERIALISABLE_RESPONSE_PROPS);
    /* eslint-enable */
    if (!response.ok) {
      throw new Error(`Response not OK: ${prettyPrint(response)}`);
    }
    this.logger.trace("Response: " + prettyPrint(response));
    return new FoodstuffsResponse(responseHandle, response);
  }

  private async isLoggedIn(page: Page) {
    type DataLayer = { loginState?: "loggedIn" | "guest" }[];
    const loginState = await page.evaluate(() => {
      const dataLayer = (globalThis as unknown as { dataLayer: DataLayer }).dataLayer;
      return dataLayer.find((entry) => "loginState" in entry)?.loginState;
    });
    return loginState === "loggedIn";
  }
}

export interface FoodstuffsUserProfile {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  agreedToMarketing: boolean;
  agreedToTermsAndConditions: boolean;
  clubCardUser: boolean;
}

export interface LoginResponse {
  success: boolean;
  userProfile: FoodstuffsUserProfile;
}

/**
 * Response backed by a Playwright JSHandle browser fetch Response. json()
 * and text() extraction methods are available.
 */
class FoodstuffsResponse implements Response {
  readonly bodyUsed: boolean;
  readonly headers: Headers;
  readonly ok: boolean;
  readonly redirected: boolean;
  readonly status: number;
  readonly statusText: string;
  readonly type: ResponseType;
  readonly url: string;

  constructor(
    private readonly responseHandle: JSHandle<Response>,
    responseJson: SerialisedResponse
  ) {
    this.bodyUsed = responseJson.bodyUsed;
    this.ok = responseJson.ok;
    this.redirected = responseJson.redirected;
    this.status = responseJson.status;
    this.statusText = responseJson.statusText;
    this.type = responseJson.type;
    this.url = responseJson.url;
    this.headers = new Headers(responseJson.headers);
  }

  /* eslint-disable */
  async json(): Promise<any> {
    const body = await this.text();
    try {
      return JSON.parse(body);
    } catch (error) {
      throw new Error("Failed to parse response body as JSON: \n" + body);
    }
  }
  /* eslint-enable */

  async text(): Promise<string> {
    return this.responseHandle.evaluate((response) => response.text());
  }

  get body(): ReadableStream<Uint8Array> | null {
    throw new Error("Cannot access body of FoodstuffsResponse directly");
  }
  get size(): number {
    throw new Error("Cannot access size of FoodstuffsResponse directly");
  }
  buffer(): Promise<Buffer> {
    throw new Error("Method not implemented.");
  }
  clone(): Response {
    throw new Error("Method not implemented.");
  }
  arrayBuffer(): Promise<ArrayBuffer> {
    throw new Error("Method not implemented.");
  }
  blob(): Promise<Blob> {
    throw new Error("Method not implemented.");
  }
  formData(): Promise<FormData> {
    throw new Error("Method not implemented.");
  }
}

const UNSERIALISABLE_RESPONSE_PROPS = [
  "arrayBuffer",
  "blob",
  "formData",
  "json",
  "text",
  "clone",
  "body",
] as const;
type UnserialisableResponseProp = typeof UNSERIALISABLE_RESPONSE_PROPS[number];

const SERIALISABLE_RESPONSE_PROPS = [
  "headers",
  "ok",
  "redirected",
  "status",
  "statusText",
  "type",
  "url",
] as const;

interface SerialisedResponse extends Omit<Response, UnserialisableResponseProp | "headers"> {
  headers: Record<string, string>;
}
