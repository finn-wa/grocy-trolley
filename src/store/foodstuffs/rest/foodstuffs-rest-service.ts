import { RestService } from "@gt/utils/rest";
import { PAKNSAVE_URL } from "../models";
import { FoodstuffsAuthHeaderProvider } from "./foodstuffs-auth-header-provider";

export abstract class FoodstuffsRestService extends RestService {
  protected readonly baseUrl = this.validateBaseUrl(`${PAKNSAVE_URL}/CommonApi`);

  constructor(protected readonly authHeaderProvider: FoodstuffsAuthHeaderProvider) {
    super();
  }

  protected authHeaders() {
    return this.authHeaderProvider.authHeaders();
  }
}
