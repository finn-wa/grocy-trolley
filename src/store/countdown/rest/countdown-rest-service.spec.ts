import { getBrowser } from "@gt/store/shared/rest/browser";
import { LoginDetails } from "@gt/store/shared/rest/login-details.model";
import { getEnvAs, initEnv } from "@gt/utils/environment";
import { Logger } from "@gt/utils/logger";
import { jest } from "@jest/globals";
import { existsSync } from "fs";
import { readdir, rm } from "fs/promises";
import * as cacheUtils from "../../../utils/cache";
import { CountdownRestService } from "./countdown-rest-service";
import { CountdownAuthHeaderProvider } from "./countdown-auth-header-provider";

class TestRestService extends CountdownRestService {
  protected readonly logger = new Logger(this.constructor.name);
  // Make authHeaders public for tests
  authHeaders = () => super.authHeaders();
}

describe("CountdownRestService", () => {
  let userAgent: CountdownAuthHeaderProvider;
  let service: TestRestService;

  initEnv({
    envFilePath: ".test.env",
    envFilePathOptional: true,
    requiredVars: ["COUNTDOWN_EMAIL", "COUNTDOWN_PASSWORD"],
  });
  const loginDetails: LoginDetails = getEnvAs({
    COUNTDOWN_EMAIL: "email",
    COUNTDOWN_PASSWORD: "password",
  });
  // Different cache dir for these tests to avoid clearing cache for other tests
  const cacheDir = cacheUtils.getCacheDirForEmail(loginDetails.email) + "_rest-service-test";
  (cacheUtils as any).getCacheDirForEmail = jest.fn((_email: string) => cacheDir);

  beforeEach(async () => {
    await rm(cacheDir, { recursive: true, force: true });
    userAgent = new CountdownAuthHeaderProvider(getBrowser, loginDetails);
    service = new TestRestService(userAgent);
  });

  afterAll(async () => {
    const browser = await getBrowser();
    await browser.close();
  });

  test("authHeaders", async () => {
    expect(existsSync(cacheDir)).toBeFalsy();
    const builder = await service.authHeaders();
    const rawHeaders = builder.raw();
    const newService = new TestRestService(userAgent);
    // should load from cache without using browser
    const cachedHeaders = (await newService.authHeaders()).raw();
    expect(cachedHeaders).toEqual(rawHeaders);
    // These assertions are more to test whether the separate cache path is working
    const cacheFiles = await readdir(`${cacheDir}/${userAgent.name}`);
    expect(cacheFiles).toMatchObject<ArrayLike<unknown>>({ length: 2 });
  });
});
