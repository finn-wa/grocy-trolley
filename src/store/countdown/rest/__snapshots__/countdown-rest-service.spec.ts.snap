// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountdownRestService getTrolley 1`] = `
Object {
  "context": Object {
    "advancedSettingsResponse": Object {
      "allowBannerRotation": true,
      "itemLevelSubstitution": true,
      "showHaveYouForgotten": true,
    },
    "basketTotals": Object {
      "bagFees": "$1.00",
      "deliveryFees": "$14.00",
      "eligibilityForDeliverySubscriptionDiscount": "NoSubscription",
      "savings": null,
      "subtotal": "$0.00",
      "totalIncludingDeliveryFees": "$14.00",
      "totalItems": 0,
    },
    "enabledFeatures": Array [
      "HaveYouForgotten",
      "ProjectShore",
      "ManagedServiceBar",
      "RecipesAddToTrolley",
      "SavedRecipes",
      "PiesEnableEachUnitForKgProducts",
      "ProductStraps",
      "DeliverySubscriptionSignUp",
      "FirstTimeShopper",
      "OrderConfirmation",
      "NewDeliverySubscriptionPage",
      "CancelOrder",
      "CountryOfOriginInAccordion",
      "ForgotPassword",
      "LazyLoadProductStampImages",
      "JQueryDeprecationLogging",
      "Optimizely",
      "ExpressDelivery",
      "ShowThatACardIsLinkedToDeliverySaver",
      "DeliverySaverBannerOnTrolleyPage",
    ],
    "fulfilment": Object {
      "address": "32 Raurenga Avenue, Royal Oak, Auckland 1023",
      "areaId": 592,
      "cutOffTime": null,
      "endTime": null,
      "expressFulfilment": Object {
        "isExpressSlot": false,
        "isLastExpressHourWindow": false,
      },
      "fulfilmentStoreId": 9538,
      "isAddressInDeliveryZone": true,
      "isDefaultDeliveryAddress": false,
      "isSlotToday": false,
      "locker": null,
      "method": "Courier",
      "perishableCode": "P",
      "selectedDate": null,
      "selectedDateWithTZInfo": null,
      "startTime": null,
    },
    "shopper": Object {
      "changingOrderId": null,
      "firstName": "Count",
      "hasActiveDeliverySubscription": false,
      "hasOnecard": false,
      "isChangingOrder": false,
      "isLoggedIn": true,
      "isPriorityShopper": false,
      "isShopper": true,
      "isSupplyLimitOverrideShopper": false,
      "oneCardBalance": Object {
        "continuitySpend": null,
        "isOneCardInError": true,
        "oneCardCurrency": "Food",
        "oneCardNumber": "",
        "onecardPointsBalance": null,
        "redeemableRewardVouchers": 0,
      },
      "orderCount": "0",
      "sessionGroups": Array [
        158,
        1961,
        1302,
        1445,
        1586,
      ],
      "shopperIdHash": "1801841f94089f01d33468e9d097f0839914c4f2753fc355398687cca595dc8b",
      "shopperScvId": "55b9e5fd-5805-4069-967b-c73deefaea84",
    },
    "shoppingListItems": Array [],
  },
  "isSuccessful": true,
  "itemCount": 0,
  "items": Array [],
  "messages": Array [],
  "rootUrl": "http://shop.countdown.co.nz",
}
`;
